File: haskell/p009/Main.hs
module Main where

solve :: Integer -> Integer
solve n = solve' n (div n 3 - 1)
  where
    solve' :: Integer -> Integer -> Integer
    solve' n a
      | c * c == a * a + b * b && b > a = (a * b) * c
      | otherwise = solve' n (a - 1)
      where
        b :: Integer
        b = div (n * (n -2 * a)) (2 * (n - a))
        c :: Integer
        c = n - a - b

main :: IO ()
main = print $ solve 1000

File: rust/p009/src/main.rs
fn solve(n: u32) -> u32 {
    // 0<a<b<c<n, a+b+c = n so a+a+a < n => 3a<n
    let (mut a, mut b, mut c) = (1, 2, 3);
    for x in (0..(n / 3 - 1)).rev() {
        a = x;
        b = (n * (n - 2 * a)) / (2 * (n - a));
        c = n - a - b;
        if b > a && c * c == b * b + a * a {
            break;
        }
    }
    a * b * c
}
fn main() {
    println!("{}", solve(1000));
}
